### ISSUE TITLE
Time Crisis "sub cpu start wait" NG (not good) on aarch64

### ISSUE THREAD

**Issue by baarreth** [mamedev/mame#8492]
When starting the game on a Raspberry Pi 4 (aarch64), the SUBCPU START WAIT test fails every time (as shown below) therefore not allowing the game to start. I tested with mame from the ubuntu 20.04 repository (0.220) and the last release compiled from the sources (0.234).

Maybe this is related to this old issue: https://mametesters.org/view.php?id=1992

mame 0.220 - ubuntu repository
![timecris_ubuntumate20 04_arm64_mame0234](https://user-images.githubusercontent.com/5972526/130847719-60660037-19b5-4a96-b26d-5356b9cc42ce.png)

mame 0.234 - compiled from sources
![timecris_ubuntumate20 04_arm64_mame0222](https://user-images.githubusercontent.com/5972526/130847715-fc150aec-0cea-45e7-adb5-2c209bff6be5.png)

This issue does not happen on the desktop.
![timecris_xubuntu20 04_x64_mame0234](https://user-images.githubusercontent.com/5972526/130847709-d8d2f67c-0cf6-4206-b27c-af06592cd4fd.png)

**Comment by angelosa** [comment]
Both references mentions using `-fsigned-char` in the build flags, anyone can test this on aarch64?

**Comment by rb6502** [comment]
macOS aarch64 is fine, but I believe the default compiler flags are x64-compatible there.

**Comment by grant2258** [comment]
I usually build everything on the pi5 with that flag so many things can cause issues without it but can test if its still relevant. Ill try get it done tommorow.

There is another issue on arm it involves unsigned int and floats, im pretty sure the mac will have this issue as well. Im not sure what the best way to fix it is though is recasting the right solution? 
More information is here https://embeddeduse.com/2013/08/25/casting-a-negative-float-to-an-unsigned-int/ not asking for a fix btw just asking if it happens on mac as well,

start eaglshot and if the gfx is messed up around the player area its caused by this bug in the v60 cpu core(recasting does fix it). I think vr has the same issues with this bug as well think it might be another cpu though would need to re-check that one.

**Comment by rb6502** [comment]
Pink garbage around the player in eaglshot?  Yeah, that does happen on the Mac.

Where are you doing the recasting to fix it?

**Comment by grant2258** [comment]
https://github.com/mamedev/mame/blob/14d11c479ecd7fbc4bc833b9d8e6366557bba399/src/devices/cpu/v60/op2.hxx#L74

**Comment by grant2258** [comment]
I looked more into the unsigned issue and there is a solution that probably will work for mac but not the pi as it only supports up till Armv8.2 extensions. You need 8.3-a and above for the fix.  

https://stackoverflow.com/questions/66279679/casting-float-to-int-with-wrap-around-on-aarch64-arm64 for an explanation of whats going on.

https://reviews.llvm.org/D64495 (add -march=armv8.4-a ) 

Sorry for going off topic on this will compile and test the original issue asked about so I answer the original question.

**Comment by happppp** [comment]
m_modwritevalw is an uint32_t. Does this work too? (coding in the same style it was)
m_modwritevalw = (uint32_t)(int32_t)val;

**Comment by cuavas** [comment]
> m_modwritevalw is an uint32_t. Does this work too? (coding in the same style it was) m_modwritevalw = (uint32_t)(int32_t)val;

But that won’t convert values between 2,147,483,648 and 4,294,967,295 properly, which I assume that code is supposed to do, considering the comment “Convert to uint32_t”.

**Comment by grant2258** [comment]
> But that won’t convert values between 2,147,483,648 and 4,294,967,295 properly, which I assume that code is supposed to do, considering the comment “Convert to uint32_t”.

so m_modwritevalw = (uint32_t)(int64_t)val; would be correct ?

**Comment by cuavas** [comment]
What’s the actual emulated CPU supposed to do for out-of-range values?  Saturate?  Trap?  Something else?

**Comment by happppp** [comment]
Yeah, not sure what would happen if there's an overflow. I'd rather @rb6502 (copyright holder of mame's v60) checks the v60 manual then.

**Comment by happppp** [comment]
![Image](https://github.com/user-attachments/assets/f55c1b3d-88aa-408c-9960-976bec708902)

CY flag unchanged (so that's a small bug in MAME)
OV flag set on integer overflow, no mention of saturation.
I don't know if the floating point exceptions are emulated in MAME.

**Comment by grant2258** [comment]
as for the original issue, the underlying cause seems to be fixed it no longer requires -fsigned-char flags to boot.

**Comment by happppp** [comment]
Thanks for checking, the issue can be closed then. Normally we'd be curious which commit fixed it, but in this case it's a lot of work to figure out.

v60/eaglshot issue: try again after this commit?
https://github.com/mamedev/mame/commit/b5fafba307ba7acc2aea90681c71a3d43aa9cac3

**Comment by grant2258** [comment]
@happppp is working fine with that commit thanks (v60/eaglshot issue)
